src/rpc/callback.rs

```diff
@@ -0,0 +1,305 @@
+            channels.push(kanal::unbounded());
+            ids.0.clone().send(i).unwrap();
+        }
+        Self {
+            callbacks: vec![std::ptr::null_mut(); REQUEST_POOL_SIZE],
+            channels,
+            ids,
+        }
+    }
+
+    pub fn init(&mut self) {
+        for i in 0..self.callbacks.len() {
+            self.callbacks[i] = Box::into_raw(Box::new(OperationCallback::new()));
+        }
+    }
+
+    pub fn free(&self) {
+        for i in 0..self.callbacks.len() {
+            unsafe {
+                drop_in_place(self.callbacks[i as usize] as *mut OperationCallback);
+            }
+        }
+    }
+
+    pub async fn register_callback(
+        &self,
+        rsp_meta_data: &mut [u8],
+        rsp_data: &mut [u8],
+    ) -> Result<u32, Box<dyn std::error::Error>> {
+        match self.ids.1.clone_async().recv().await {
+            Ok(id) => {
+                let callback =
+                    unsafe { &mut *(self.callbacks[id as usize] as *mut OperationCallback) };
+                callback.data = rsp_data.as_ptr();
+                callback.meta_data = rsp_meta_data.as_ptr();
+                Ok(id)
+            }
+            Err(_) => Err("register_callback error")?,
+        }
+    }
+
+    pub fn get_data_ref(
+        &self,
+        id: u32,
+        data_length: usize,
+    ) -> Result<&mut [u8], Box<dyn std::error::Error>> {
+        let callback = self.callbacks[id as usize];
+        unsafe {
+            Ok(std::slice::from_raw_parts_mut(
+                (*callback).data as *mut u8,
+                data_length,
+            ))
+        }
+    }
+
+    pub fn get_meta_data_ref(
+        &self,
+        id: u32,
+        meta_data_length: usize,
+    ) -> Result<&mut [u8], Box<dyn std::error::Error>> {
+        let callback = self.callbacks[id as usize];
+        unsafe {
+            Ok(std::slice::from_raw_parts_mut(
+                (*callback).meta_data as *mut u8,
+                meta_data_length,
+            ))
+        }
+    }
+
+    pub async fn response(
+        &self,
+        id: u32,
+        status: libc::c_int,
+        flags: u32,
+        meta_data_length: usize,
+        data_lenght: usize,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        {
+            let callback = unsafe { &mut *(self.callbacks[id as usize] as *mut OperationCallback) };
+            callback.request_status = status;
+            callback.flags = flags;
+            callback.meta_data_length = meta_data_length;
+            callback.data_length = data_lenght;
+        }
+        self.channels[id as usize].0.clone_async().send(()).await?;
+        Ok(())
+    }
+
+    // wait_for_callback
+    // return: (status, flags, meta_data_length, data_length)
+    pub async fn wait_for_callback(
+        &self,
+        id: u32,
+    ) -> Result<(i32, u32, usize, usize), Box<dyn std::error::Error>> {
+        let result = self.channels[id as usize].1.clone_async().recv().await;
```

Comment:
Oh, you use channels as mpmc models. This is a good temporary work, but we can update them to mpmc models, which are lock free, in the future.